### PR TITLE
implement logical plan.

### PR DESCRIPTION
**Pull request by hanfei19910905**
Finally, the plan will be specified to LogicalPlan and physical plan. This patch implements logic plan first.
@shenli @coocood @zimulala  PTAL

### CODE REVIEW

File: plan/plan.go [hunk 1]
```diff
@@ -62,6 +96,60 @@ type Plan interface {
...
+
+	// PruneColumnsAndResolveIndices prunes unused columns and resolves index for columns.
+	// This function returns a column slice representing outer columns and an error.
+	PruneColumnsAndResolveIndices([]*expression.Column) ([]*expression.Column, error)
+	// TODO: implement Convert2PhysicalPlan()
+}
+
+// TODO: implement PhysicalPlan
+
+type logicalPlan struct {
```
Comment (coocood) [inline plan/plan.go]: rename it to `baseLogicalPlan` is better.

File: plan/newplanbuilder.go [hunk 2]
```diff
@@ -27,14 +27,20 @@ import (
 // UseNewPlanner means if use the new planner.
 var UseNewPlanner = false
 
-func (b *planBuilder) allocID(p Plan) string {
-	b.id++
-	return fmt.Sprintf("%T_%d", p, b.id)
+type idAllocer struct {
```
Comment (coocood) [inline plan/newplanbuilder.go]: rename to 'idAllocator' is better.

File: plan/planbuilder.go [hunk 3]
```diff
@@ -58,7 +58,7 @@ type planBuilder struct {
 	hasAgg       bool
 	sb           SubQueryBuilder
 	obj          interface{}
-	id           int
+	allocer      *idAllocer
```
Comment (coocood) [inline plan/planbuilder.go]: rename the field to 'allocator' is better.

File: plan/plan.go [hunk 4]
```diff
@@ -62,6 +96,60 @@ type Plan interface {
 	IsCorrelated() bool
 }
 
+// LogicalPlan is a tree of logical operators.
+// We can do a lot of logical optimization to it, like predicate push down and column pruning.
+type LogicalPlan interface {
+	Plan
+
+	PredicatePushDown([]expression.Expression) ([]expression.Expression, error)
```
Comment (shenli) [inline plan/plan.go]: Add comment for this function.

File: plan/plan.go [hunk 5]
```diff
@@ -111,27 +111,27 @@ type LogicalPlan interface {
...
 // TODO: implement PhysicalPlan
 
-type logicalPlan struct {
+type baseLogicalPlan struct {
 	basePlan
 }
 
-func newLogicalPlan(tp string, a *idAllocer) logicalPlan {
-	return logicalPlan{
+func newLogicalPlan(tp string, a *idAllocator) baseLogicalPlan {
```
Comment (coocood) [inline plan/plan.go]: newBaseLogicalPlan

**Comment by shenli** [comment]
LGTM

**Comment by coocood** [comment]
LGTM

**Pull request merged**